core/src/main/java/kafka/metrics/ClientMetricsConfigs.java

```diff
@@ -0,0 +1,194 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.metrics;
+
+import org.apache.kafka.common.config.AbstractConfig;
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.common.config.ConfigDef.Importance;
+import org.apache.kafka.common.config.ConfigDef.Type;
+import org.apache.kafka.common.errors.InvalidConfigurationException;
+import org.apache.kafka.common.errors.InvalidRequestException;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+/**
+ * Client metric configuration related parameters and the supporting methods like validation, etc. are
+ * defined in this class.
+ * <p>
+ * {
+ * <ul>
+ *   <li> name: Name supplied by CLI during the creation of the client metric subscription.
+ *   <li> metrics: List of metric prefixes
+ *   <li> intervalMs: A positive integer value >=0  tells the client that how often a client can push the metrics
+ *   <li> match: List of client matching patterns, that are used by broker to match the client instance
+ *   with the subscription.
+ * </ul>
+ * }
+ * <p>
+ * At present, CLI can pass the following parameters in request to add/delete/update the client metrics
+ * subscription:
+ * <ul>
+ *    <li> "name" is a unique name for the subscription. This is used to identify the subscription in
+ *          the broker. Ex: "METRICS-SUB"
+ *    <li> "metrics" value should be comma separated metrics list. A prefix match on the requested metrics
+ *          is performed in clients to determine subscribed metrics. An empty list means no metrics subscribed.
+ *          A list containing just an empty string means all metrics subscribed.
+ *          Ex: "org.apache.kafka.producer.partition.queue.,org.apache.kafka.producer.partition.latency"
+ *
+ *    <li> "interval.ms" should be between 100 and 3600000 (1 hour). This is the interval at which the client
+ *          should push the metrics to the broker.
+ *
+ *    <li> "match" is a comma separated list of client match patterns, in case if there is no matching
+ *          pattern specified then broker considers that as all match which means the associated metrics
+ *          applies to all the clients. Ex: "client_software_name = Java, client_software_version = 11.1.*"
+ *          which means all Java clients with any sub versions of 11.1 will be matched i.e. 11.1.1, 11.1.2 etc.
+ * </ul>
+ * For more information please look at kip-714:
+ * https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability#KIP714:Clientmetricsandobservability-Clientmetricsconfiguration
+ */
+public class ClientMetricsConfigs extends AbstractConfig {
```

Comment:
This class is major overlap from PR where it was introduced: https://github.com/apache/kafka/pull/14621

Reply:
Resolved conflicts after merging the upstream.